src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java

```diff
@@ -172,65 +198,331 @@ private void handleError(@Nonnull final String errorMessage, @Nullable final Thr
+     * batch (maybe with a different reference or a reference that doesn't exist anymore).
      *
-     * @param productTypeDraft the product type draft to create the product type from.
-     * @return a {@link CompletionStage} which contains an empty result after execution of the create.
+     *
+     * <p>Then, makes a copy of {@code productTypeDraft} and then goes through all its attribute definition drafts.
+     * For each attribute, attempts to find a nested product type in its attribute type, if it has a
+     * nested type. It checks if the key, of the productType reference, is cached in {@code keyToIdCache}. If it is,
+     * then it means the referenced product type exists in the target project, so there is no need to remove or keep
+     * track of it. If it is not, it means it doesn't exist yet and needs to be tracked as a
+     * missing reference and also remove this attribute definition from the supplied {@code draftCopy} to be able to
+     * create product type without this attribute containing the missing reference.
+     *
+     *
+     * @param productTypeDraft         the productTypeDraft containing the attribute which should be updated by removing
+     *                                 the attribute which contains the missing reference.
+     * @param keyToIdCache             a map of productType key to id. It represents a cache of the existing
+     *                                 productTypes in the target project.
      */
+    @SuppressWarnings("ConstantConditions") // since the batch is validate before, key is assured to be non-blank here.
     @Nonnull
-    private CompletionStage<Optional<ProductType>> applyCallbackAndCreate(
-        @Nonnull final ProductTypeDraft productTypeDraft) {
+    private ProductTypeDraft removeAndKeepTrackOfMissingNestedAttributes(
+        @Nonnull final ProductTypeDraft productTypeDraft,
+        @Nonnull final Map<String, String> keyToIdCache) {
 
-        return syncOptions
-            .applyBeforeCreateCallBack(productTypeDraft)
-            .map(draft -> productTypeService
-                .createProductType(draft)
-                .thenApply(productTypeOptional -> {
-                    if (productTypeOptional.isPresent()) {
-                        statistics.incrementCreated();
-                    } else {
-                        statistics.incrementFailed();
+        statistics.removeReferencingProductTypeKey(productTypeDraft.getKey());
+
+        final List<AttributeDefinitionDraft> attributeDefinitionDrafts = productTypeDraft.getAttributes();
+        if (attributeDefinitionDrafts == null || attributeDefinitionDrafts.isEmpty()) {
+            return productTypeDraft;
+        }
+
+        // copies to avoid mutation of attributes array supplied by user.
+        final ProductTypeDraft draftCopy = ProductTypeDraftBuilder
+            .of(productTypeDraft)
+            .attributes(new ArrayList<>(productTypeDraft.getAttributes()))
```

Comment:
Are you sure this is a deep copy? Seems to me you're still referencing the attributes of the original product type draft.

Reply:
yes, this copies the list but doesn't copy the actual attributes. And I only remove from the list but not manipulate the actual elements.

Check the implementation of this method: https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#ArrayList-java.util.Collection-

it copies.